libraries/hotstuff/include/eosio/hotstuff/qc_chain.hpp

```diff
@@ -77,71 +77,261 @@ namespace eosio::hotstuff {
+
+      bool                 is_quorum_met() const { return valid(); }
+      const fc::sha256&    get_proposal_id() const { return _proposal_id; }
+      std::string          get_votes_string() const {
+         return std::string("strong(\"") + bitset_to_string(_strong_votes._bitset) + "\", weak(\"" +
+            bitset_to_string(_weak_votes._bitset) + "\"";
+      }
+      // ================== end compatibility functions =======================
+
+      void reset(const fc::sha256& proposal_id, const digest_type& proposal_digest, size_t num_finalizers, size_t quorum) {
+         _proposal_id = proposal_id;
+         _proposal_digest.assign(proposal_digest.data(), proposal_digest.data() + 32);
+         _quorum = quorum;
+         _strong_votes.reset(num_finalizers);
+         _weak_votes.reset(num_finalizers);
+         _num_finalizers = num_finalizers;
+         _state = state_t::unrestricted;
       }
 
-      void reset(const fc::sha256& proposal, size_t finalizer_size) {
-         proposal_id = proposal;
-         active_finalizers = hs_bitset{finalizer_size};
-         active_agg_sig = fc::crypto::blslib::bls_signature();
-         quorum_met = false;
+      bool add_strong_vote(const std::vector<uint8_t>& proposal_digest,
+                           size_t index,
+                           const bls_public_key& pubkey,
+                           const bls_signature& sig) {
+         assert(index < _num_finalizers);
+         if (!_strong_votes.add_vote(proposal_digest, index, pubkey, sig))
+            return false;
+         size_t weak   = num_weak();
+         size_t strong = num_strong();
+         
+         switch(_state) {
+         case state_t::unrestricted:
+         case state_t::restricted:
+            if (strong >= _quorum)
+                _state = state_t::strong;
+            else if (weak + strong >= _quorum)
+               _state = state_t::weak_achieved;
+            break;
+            
+         case state_t::weak_achieved:
+            if (strong >= _quorum)
+               _state = state_t::strong;
+            break;
+            
+         case state_t::weak_final:
+         case state_t::strong:
+            // getting another strong vote...nothing to do
+            break;
+         }
+         return true;
       }
 
-      const hs_bitset& get_active_finalizers() const {
-         assert(!active_finalizers.empty());
-         return active_finalizers;
+      bool add_weak_vote(const std::vector<uint8_t>& proposal_digest,
+                         size_t index,
+                         const bls_public_key& pubkey,
+                         const bls_signature& sig) {
+         assert(index < _num_finalizers);
+         if (!_weak_votes.add_vote(proposal_digest, index, pubkey, sig))
+            return false;
+         size_t weak   = num_weak();
+         size_t strong = num_strong();
+         
+         switch(_state) {
+         case state_t::unrestricted:
+         case state_t::restricted:
```

Comment:
`restricted` should be moved out of here and moved down.

Reply:
Hum, I don't think so.